Change rock.conv2d to rock.conv, change y/x and h/w to 0/1 in layout keys, make some dimension scalars into arrays

pcf000 opened the pull request:
Making a pull request for sanity check as much as for potential merging.

Changed "conv2d" to "conv" nearly everywhere, including the --operation flag to rocmlir-gen.  That's a set of straightforward diffs in the first commit.

Changing y/x and h/w to 0/1 and making some scalars into vectors are intermingled changes and spread over several commits, along with additional changes to make tests work.  The letters in the layout are used in several ways, and converted back and forth from layout strings to layout arrays to layout string-maps. I've tried to not affect outwardly-visible keys too much, which means (a) there are about four places that accept y/x/h/w and convert it to 0/1 and (b) the changes in tests are mostly in comparison strings and not in test commands.

I'll have a fair number of comments in the changes.  There are some places I want to highlight because they're either incomplete, will require further work to extend to 3-D, or raise questions for me.

pcf000 committed:
Changing 'rock.conv2d' to 'rock.conv'.

pcf000 committed:
Behind-the-scenes changes, mostly related to ConvolutionDims.

pcf000 committed:
Rebase glitch.

pcf000 committed:
Convert x/y to 1/0 in ConvToGemm.cpp, though still no loops, and some affected tests.

pcf000 committed:
Make strideH/W into an array, first step to doing dilation, padding, 'tilda', etc.

pcf000 committed:
Making arrays out of some dilation and stride variables, and dependents.

pcf000 committed:
Making arrays out of padding variables and dependents.

pcf000 committed:
Cleanup.

pcf000 committed:
Methodical test fixing.  Still fails the two resnet18 pr-e2e xmir tests.  Some debug code still present.

pcf000 committed:
To fix the two resnet18_blk* tests, have both hw and 01 entries in the match-layouts-to-input tables.

pcf000 committed:
Remove some debug code.

pcf000 commented on mlir/include/mlir/Dialect/Rock/Generator/ConvGenerator.h:
```diff
@@ -130,6 +130,13 @@ class ConvGenerator {
                    std::back_inserter(targetLayout),
                    [&targetSpec](int64_t p) { return targetSpec[p]; });
+
+    // +++pf:  update old key names.
```
I used this same comment in all the places I have this kind of conversion code.  My intent is that we'll remove it as we finish the changeover.

pcf000 commented on mlir/include/mlir/Dialect/Rock/IR/ConvolutionDims.h:
```diff
@@ -23,20 +25,19 @@ namespace rock {
-  int64_t wo;
-  int64_t hi;
-  int64_t wi;
+  SmallVector<int64_t, 5> fil;
```
These guys become arrays, which affects both how ConvolutionDims is constructed by its users and how the data is extracted and passed on from the struct.

krzysz00 replied:
I think you'll only need 3 or 4 here, since n, k, c, and g are held separately

pcf000 commented on mlir/include/mlir/Dialect/Rock/IR/RockAttrDefs.td:
```diff
@@ -50,23 +50,23 @@ class Rock_Attr<string name, list<Trait> traits = []> :
-def ConvOp2DType : I32EnumAttrCase<"Fwd", 0, "conv2d">;
-def ConvOp2DBwdDataType : I32EnumAttrCase<"BwdData", 1, "conv2d_bwd_data">;
-def ConvOp2DBwdWeightType : I32EnumAttrCase<"BwdWeight", 2, "conv2d_bwd_weight">;
+def ConvOpType : I32EnumAttrCase<"Fwd", 0, "conv">;
```
The heart of the conv2d-to-conv name change.  I could locate most of the ripples by making this change, compiling, and seeing what broke.

pcf000 commented on mlir/include/mlir/Dialect/Rock/Tuning/GridwiseGemmParams.h:
```diff
@@ -194,10 +194,7 @@ class BasePopulateParams {
-          return true;
-        }
-        return false;
+        return (this->original_pos < rhs.original_pos);
```
clangd pointed this one out to me.

pcf000 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -175,27 +173,14 @@ makeRockConv2D(ConversionPatternRewriter &rw, Operation *op, Value input,
   rock::GemmFeatures features;
   std::tie(arch, num_cu, features) = getArchAttributes(op, input.getType());
 
-  ArrayRef<int64_t> pad64 = pad;
```
If we're getting an attribute array, we don't need to make intermediate arrays, to make another attribute array.

pcf000 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -186,9 +186,9 @@ LogicalResult ConvGenerator::hasValidDimension() const {
-  if (failed(hasDimensions(inDim, "ngchw", "input")) ||
-      failed(hasDimensions(filDim, "gkcyx", "filter")) ||
-      failed(hasDimensions(outDim, "ngkhw", "output"))) {
+  if (failed(hasDimensions(inDim, "ngc01", "input")) ||
```
First update of layout strings.  Can't have something like "ngchw01" because the test cares about the length.

pcf000 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -215,37 +215,37 @@ LogicalResult ConvGenerator::hasValidDimension() const {
 
   int64_t expectedOutHeight = outputDim(
-      inDim["h"], filDim["y"], config.paddingLeftDims[DIM::HEIGHT],
+      inDim["0"], filDim["0"], config.paddingLeftDims[DIM::HEIGHT],
```
inDim and filDim are StringMaps;  the key strings change.  DIM::HEIGHT is part of an enum defined in ConvGenerator.h, because I didn't want to hard-code 0 and 1 everywhere.

krzysz00 replied:
I think these are all loop-computable, subject to the awkward need to `formatv` or `Twine()` your loop variable.

pcf000 replied:
Commit 4d3dff009 does that, along with some places where, eg, llvm::any_of() is more concise than std::any_of().

pcf000 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -654,30 +654,38 @@ ConvGenerator::parseConvDims(int64_t batchSize, int64_t groupSize,
                              int64_t filterHeight, int64_t filterWidth) {
   config.filterDims[DIM::HEIGHT] = filterHeight;
   config.filterDims[DIM::WIDTH] = filterWidth;
-  static const std::string filterKeys = "kgcyx";
```
The logic here bothered me.  Keys and Vals are parallel arrays -- given a layout key character, we'll find() it in the Keys array (string) to get an index, and use that index to look up the value in Vals.  Why not just a StringMap to associate the key and value directly?

pcf000 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -654,30 +654,38 @@ ConvGenerator::parseConvDims(int64_t batchSize, int64_t groupSize,
+                                        {"c", inputChannel / groupSize},
+                                        {"y", filterHeight},
+                                        {"x", filterWidth},
+                                        {"0", filterHeight},
```
Both old and new keys are present in the map, for now.

krzysz00 replied:
This is a nice cleanup

pcf000 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -421,8 +421,12 @@ ConvolutionDims ConvolutionDims::fromOp(Operation *op) {
 
     if (filterAttr.getValue() == "y") {
       y = filterShape[i];
+    } else if (filterAttr.getValue() == "0") {
```
Another place that accepts both old and new keys.  I thought I could just convert old forms to new at a few entry points, but then there are places that read conv ops from a file and there's no place to convert before they call functions like fromOp().  It surprised me how much of Operation construction is completely generic.

pcf000 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -443,19 +451,23 @@ ConvolutionDims ConvolutionDims::fromOp(Operation *op) {
   }
 
-  return ConvolutionDims(y, x, ho, wo, hi, wi, k, c, n, g);
+  return ConvolutionDims({y, x}, {ho, wo}, {hi, wi}, k, c, n, g);
```
Because ConvolutionDims has arrays for those dimensions, we can aggregate them in the constructor call here.

The logic here will need brushing up to extend to 3-D.

pcf000 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -489,14 +501,14 @@ GemmSize GemmSize::fromConvolution(ConvOpType type,
     gemmMSize = sizes.k;
-    gemmKSize = sizes.c * sizes.y * sizes.x;
-    gemmNSize = sizes.n * sizes.ho * sizes.wo;
+    gemmKSize = sizes.c * sizes.fil[0] * sizes.fil[1];
```
References to arrays now.  Not sure how some of these expressions should be extended for 3-D.

krzysz00 replied:
These expressions extend naturally - they're the product of the array.

pcf000 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -555,8 +567,8 @@ static LogicalResult verifyConvOp(RockConvInterface convOp) {
 
-  if (isDisjointed("filter_layout", "y", "x") ||
-      isDisjointed("input_layout", "hi", "wi"))
+  if ((isDisjointed("filter_layout", "y", "x") && isDisjointed("filter_layout", "0", "1")) ||
```
Accepting both old and new.

krzysz00 replied:
Is this correct? It feels like the arms of the `&&` are mutually exclusive

pcf000 commented on mlir/lib/Dialect/Rock/IR/RockDialect.cpp:
```diff
@@ -655,10 +667,10 @@ GemmSize Conv2DBwdDataOp::getGemmSize() {
   int64_t yTilda = strideH / gcdStrideDilationH;
```
Have not turned some of these variables into arrays and thus have no loops for the dimensions.

krzysz00 replied:
These _should_ generalize out nicely? But that's a future PR, I think

krzysz00 replied:
And, right, this is a different file and we're probably looking at a struct that could be unified away

pcf000 commented on mlir/lib/Dialect/Rock/IR/TransformMapBuilder.cpp:
```diff
@@ -260,7 +260,10 @@ StringRef TransformMapBuilder::endName(uint32_t dim) {
 uint32_t TransformMapBuilder::startIndex(StringRef name) {
-  assert(startIndices.count(name) == 1 && "Key not in starting set of names");
+  if (startIndices.count(name) != 1) {
+    llvm::errs() << "Key not in starting set of names: " << name << "\n";
```
I wanted to see *which* key wasn't in the set.

krzysz00 replied:
I think we've got `report_fatal_error` down in LLVM for this sort of thing?

pcf000 commented on mlir/lib/Dialect/Rock/Tuning/ConvContext.cpp:
```diff
@@ -23,15 +23,26 @@ populateDimIndexAndSize(const ArrayAttr &layoutAttr,
+    if (key == "y") key = "0";
+    if (key == "x") key = "1";
+    if (key[0] == 'h')
+      key = StringAttr::get(layoutAttr.getContext(),
```
It is a pain in the butt to take a string attribute and replace a character.

krzysz00 replied:
You shouldn't need to `std::string` here. `Twine()` should do you?

pcf000 commented on mlir/lib/Dialect/Rock/utility/loweringUtils.cpp:
```diff
@@ -66,6 +66,9 @@ LogicalResult mlir::rock::calculateKBlockNum(const int64_t batchSize,
 
   int64_t gemmKBlock = 1;
 
+  assert(gemmM > 0 && gemmN > 0 && gemmK > 0);
```
One common failure mode was to divide by zero down here.

pcf000 commented on mlir/test/Dialect/Rock/conv_to_gemm_layout_regularization.mlir:
```diff
@@ -1,20 +1,20 @@
-// RUN: rocmlir-gen --arch gfx908 -p -operation conv2d -fil_layout gkyxc -in_layout ngchw -out_layout hwgkn \
+// RUN: rocmlir-gen --arch gfx908 -p -operation conv -fil_layout gkyxc -in_layout ngchw -out_layout hwgkn \
```
Changed the operation but not the layout, but changed the expected layout keys in the transform maps.  A bit inconsistent.

pcf000 commented on mlir/tools/rocmlir-gen/rocmlir-gen.cpp:
```diff
@@ -793,6 +793,14 @@ static void correctConvParameters() {
   else if (inputLayoutValue.size() == 4)
     inputLayout = "g" + inputLayoutValue;
 
+  // +++pf:  update old key names.
```
rocmlir-gen will accept either old or new keys, for now.

krzysz00 replied:
This we might as well keep around for ergonomics

pcf000 commented on mlir/tools/rocmlir-gen/rocmlir-gen.cpp:
```diff
@@ -1767,24 +1775,53 @@ createCPUConvWithMLIR(ModuleOp module, func::FuncOp &func,
              direction == rock::ConvOpType::BwdWeight) &&
             tensor == INPUT) {
-          if (c == 'h') {
+          if (c == '0') {               // +++pf: may need adjustment to h/w.
```
This whole section needs to be reworked.  It's connecting loopIVs by finding the right letter in the layout string, and needs to have both y/x and h/w present, which causes a collision since 0/1 could mean either.  At the moment, the string will have both 0/1 and y/x, but the better solution is to use a vector of strings like "0i" to allow for unambiguous keys.

There's also logic to override the key match in certain cases, which is necessary to make some test cases work but is confused a bit by the key renaming.  The set of changes here works but is not the final answer.

pcf000 commented on mlir/tools/rocmlir-gen/rocmlir-gen.cpp:
```diff
@@ -2000,6 +2037,8 @@ static void createCPUConvWithCPP(ModuleOp module, func::FuncOp &func,
   layoutConstOps['n'] = nConstantOp;
   layoutConstOps['h'] = hConstantOp;
   layoutConstOps['w'] = wConstantOp;
+  layoutConstOps['0'] = yConstantOp;
```
Need to accept both old and new keys.

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -87,6 +87,12 @@ template <typename T>
+//   printf("checkNames({");
+//   for (auto name : actual) printf("%s,", name);
+//   printf("}, {");
+//   for (auto name : expected) printf("%s,", name);
```
Stray comments (that could be `LLVM_DEBUG()`?)

pcf000 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -87,6 +87,12 @@ template <typename T>
 LogicalResult checkNames(ArrayRef<StringRef> actual,
                          ArrayRef<StringRef> expected, StringRef argName,
                          T op) {
+//   printf("checkNames({");
```
Oops, debug code remaining.

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -134,22 +140,41 @@ LogicalResult getConvDimNames(T op, SmallVectorImpl<StringRef> &filterNames,
+                           if (name == "y") return StringAttr::get(ctx, "0");
+                           if (name == "x") return StringAttr::get(ctx, "1");
+                           if (name.strref().starts_with_insensitive("h")) {
+                             auto namestr = name.str();
```
Do you need the `.str()` here. And does that return a StringRef or? 

(and, on the third hand, does the string attr constructor take `Twine`?)

pcf000 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -474,10 +499,14 @@ struct MatchLayoutsToInput final
     TypedValue<ShapedType> filter = op.getFilter(), output = op.getOutput();
     const llvm::StringMap<StringAttr> inputToFilter = {
         {"ci", b.getStringAttr("c")},
+        {"0i", b.getStringAttr("0")},
```
Need both old and new forms.  Without this change, resnet18_blk.mlir has an RMS of 1e-3 instead of 1e-7.

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -598,7 +621,7 @@ LogicalResult backwardWeightAtomicAdd(Conv2DBwdWeightOp op,
     llvm::StringMap<uint32_t> firstTransformOutDims = expandNamesInPlace(
         inputNames,
-        {{"ni", {"n0", "n1"}}, {"hi", {"hipad"}}, {"wi", {"wipad"}}});
+        {{"ni", {"n0", "n1"}}, {"0i", {"hipad"}}, {"1i", {"wipad"}}});
```
We'll likely want to refactor here so we get `0i_pad` , `1i_pad` and so on.

pcf000 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -534,16 +563,10 @@ LogicalResult backwardWeightAtomicAdd(Conv2DBwdWeightOp op,
-  int64_t strideH = ctx.getStrideVal()[0];
-  int64_t strideW = ctx.getStrideVal()[1];
+  // Obtain convolution parameters: padding / dilation / stride.
+  auto pads = ctx.getPaddingVal();
```
Intermediate arrays from the arrays in ConvolutionDims.

pcf000 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -661,7 +683,7 @@ LogicalResult backwardWeightAtomicAdd(Conv2DBwdWeightOp op,
     firstWrap.unmerge({"n0", "n1"}, "no",
                       {gemmKBlocks, convDims.n / gemmKBlocks});
-    firstWrap.passThrough({"ko", "ho", "wo"});
+    firstWrap.passThrough({"ko", "0o", "1o"});
```
How does something like translate to 1-D and 3-D?

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -635,9 +657,9 @@ LogicalResult backwardWeightAtomicAdd(Conv2DBwdWeightOp op,
+    llvm::SmallVector<StringRef, 3> nonNHWDims = {"ci", "0", "1"};
     matchUnderlyingOrder(nonNHWDims, gemmInputTransform);
-    llvm::SmallVector<StringRef, 3> nhwDims = {"n1", "ho", "wo"};
+    llvm::SmallVector<StringRef, 3> nhwDims = {"n1", "0o", "1o"};
```
Future work will be generating this list with a `for (i : numberOfImageDimensions)` and some calls to `Twine`.

... Heck, while we're here, we might go ahead and extend `TransformMapBuilder` to let it work on `StringAttr`s transparently.

pcf000 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -775,15 +797,15 @@ LogicalResult backwardData(Conv2DBwdDataOp op, PatternRewriter &b) {
     // particular embed coefficients is in a presentation somewhere)
     llvm::StringMap<uint32_t> embedDims = expandNamesInPlace(
-        filterNames, {{"y", {"ydot", "ytilda"}}, {"x", {"xdot", "xtilda"}}});
+        filterNames, {{"0", {"0dot", "0tilda"}}, {"1", {"1dot", "1tilda"}}});
```
Need to also make a loop that generates the "Ndot" and "Ntilda" strings, and can then make the tilda and slice scalars into arrays and make this section into loops.

krzysz00 commented on mlir/test/Conversion/RockToGPU/multiple_kernels.mlir:
```diff
@@ -13,22 +13,22 @@
 module  {
   func.func @step1(%arg0: memref<1x512x16x3x3xf32>, %arg1: memref<512x1x16x32x32xf32>, %arg2: memref<512x1x512x30x30xf32>) attributes {kernel, arch = "amdgcn-amd-amdhsa:gfx906"} {
-    rock.conv2d(%arg0, %arg1, %arg2) features = dot {arch = "amdgcn-amd-amdhsa:gfx906", dilations = [1 : index, 1 : index], filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], output_layout = ["no", "go", "ko", "ho", "wo"], padding = [0 : index, 0 : index, 0 : index, 0 : index], strides = [1 : index, 1 : index]} : memref<1x512x16x3x3xf32>, memref<512x1x16x32x ...
+    rock.conv(%arg0, %arg1, %arg2) features = dot {arch = "amdgcn-amd-amdhsa:gfx906", dilations = [1 : index, 1 : index], filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], output_layout = ["no", "go", "ko", "ho", "wo"], padding = [0 : index, 0 : index, 0 : index, 0 : index], strides = [1 : index, 1 : index]} : memref<1x512x16x3x3xf32>, memref<512x1x16x32x32 ...
```
Let's go ahead and output the quoted strings here? Just so we have canonical-form IR in our tests.

pcf000 replied:
Not sure what you mean by "output the quoted strings";  something like "ngchw" instead of the list of strings?

krzysz00 commented on mlir/test/Conversion/TosaToRock/no-mixed-layout.mlir:
```diff
@@ -1,6 +1,6 @@
 // RUN: rocmlir-driver -host-pipeline partition,highlevel -targets %arch %s | FileCheck %s
 
-// CHECK: rock.conv2d({{.*}}) {{.*}} {{{.*}}, filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], output_layout = ["no", "go", "ko", "ho", "wo"]{{.*}}}
+// CHECK: rock.conv({{.*}}) {{.*}} {{{.*}}, filter_layout = ["g", "k", "c", "y", "x"], input_layout = ["ni", "gi", "ci", "hi", "wi"], output_layout = ["no", "go", "ko", "ho", "wo"]{{.*}}}
```
Is updating TosaToRock a future PR?

pcf000 replied:
I consider non-incidental Tosa stuff to be part of issue #1314.

krzysz00 reviewed (commented):
Overall, seems fine, though I'd start to hitting our tests with the big search-replace hammer

pcf000 committed:
Restore missing files.

pcf000 committed:
Merge branch 'develop' into pf-issue-1311c

pcf000 commented:
Should I do the big test search-and-replace for this pull request?

pcf000 committed:
Also needed to change conv2d to conv in tuning code.

krzysz00 commented:
@pcf000 I'd prefer if we did

pcf000 committed:
clang-format

pcf000 committed:
Merge branch 'develop' into pf-issue-1311c

pcf000 committed:
clang-format again.

pcf000 committed:
Change hw to 01 for pad, tilda, slice.  Use 0f and 0i when both y and h are being used.

pcf000 committed:
Finish converting scalars to vectors and loops in ConvToGemm.cpp
For expanding names, make strings that can be aggregated into a
StringMap for passing as an initialisation.  For most other things,
either loop doing the operation for the given dimension index, or loop
across all dimensions to accumulate a list of strings to pass to a
transform function.
All created strings are made with b.getStringAttr() so they stick around
across iterations.

pcf000 committed:
Tiny bug fix.

pcf000 committed:
Big search-and-replace of y/x/h/w to 0/1, except for seven Tosa tests that'll come with #1314.

pcf000 committed:
Generalise backwardData()'s delinearisation of kernelId, though only 3-D is added.  Small cleanups.

pcf000 committed:
Convert getGemmSize() and backwardDataKernelIds() to use loops over N dimensions.

pcf000 committed:
Fix bug in getKernelCount() array version.

krzysz00 commented on mlir/lib/Conversion/TosaToRock/TosaToRock.cpp:
```diff
@@ -111,7 +111,7 @@ static Value expandTensor(ConversionPatternRewriter &rw, Operation *op,
 getArchAttributes(Operation *op, Type inputType) {
   auto func = op->getParentOfType<func::FuncOp>();
-  auto mod = func->getParentOfType<ModuleOp>();
+  //auto mod = func->getParentOfType<ModuleOp>();
```
... this shouldn't be dead code, I think we can still have cases where the module is `arch`'d and the function isn't?

pcf000 replied:
I didn't write it, I just got tired of the unused-variable warning.

krzysz00 commented on mlir/lib/Dialect/Rock/utility/loweringUtils.cpp:
```diff
@@ -132,18 +133,42 @@ mlir::rock::backwardDataKernelIds(int64_t strideHeight, int64_t strideWidth,
+      [[fallthrough]];
+    case 2:
+      iTilda[1] = (kernelId % subproduct) / divisor;
+      iTilda[0] = kernelId / subproduct;
```
I think this can be generalized to N dimensions. 

... heck, you probably want `delinearize` from `external/llvm-project/mlir/include/mlir/Dialect/Utils/IndexingUtils.h`
, which makes it `iTilda = delinearize(kernelID, filDims);`

pcf000 replied:
The delinearize() function is computing something slightly different, which is giving me different results in practice.  Could be I need to call it differently, or could be that our needs just differ.

krzysz00 replied:
Could be, but yeah, this is a loop

```
SmallVector<int64_t> iTilda(0, filterDims.size());
int64_t remainingKernelId = kernelId;
for (auto [thisITilda, dim] : llvm::zip_equal(llvm::reverse(iTilda, filterDims)) {
  thisITilda = remainingKernelId % dim;
  remainingKernelId /= dim;
}
```

krzysz00 replied:
Which should be exactly `delinearize(kernelId, filterDims);`

krzysz00 commented on mlir/test/Dialect/Rock/integration/reduce/blockwise_reduce/blockwise_reducemax_nr_threads_lt_blocksize.mlir:
```diff
@@ -1,11 +1,11 @@
 // RUN: rocmlir-gen -ph -print-results -rand fixed - < %s | rocmlir-driver -arch %arch -c  | mlir-cpu-runner -O2 --shared-libs=%linalg_test_lib_dir/libmlir_rocm_runtime%shlibext,%conv_validation_wrapper_library_dir/libconv-validation-wrappers%shlibext,%linalg_test_lib_dir/libmlir_runner_utils%shlibext,%linalg_test_lib_dir/libmlir_float16_utils%shlibext --entry-point-result=void | FileCheck %s
```
I think you hit an overzealous sed on this file and can revert

krzysz00 commented on mlir/lib/Dialect/Rock/Transforms/ConvToGemm.cpp:
```diff
@@ -607,25 +634,41 @@ LogicalResult backwardWeightAtomicAdd(Conv2DBwdWeightOp op,
+      StringAttr key = b.getStringAttr(Twine(i) + "ipad");
+      StringAttr val1 = b.getStringAttr(Twine(i));
+      StringAttr val2 = b.getStringAttr(Twine(i) + "o");
+      expansions.insert({key, {val1, val2}});
```
Huh, does `StringMap` not own its keys?

pcf000 replied:
Nope, it maps StringRefs to things.  I even witnessed strings changing out from under the StringMap.

krzysz00 reviewed (commented):
This here's coming along well (haven't really stared at all the exact details but they're probably good).

Do let me know when you think we've hit a point where hitting merge and moving to a followup PR is reasonable.

pcf000 committed:
clang-format

pcf000 committed:
Convert hasValidDimension() to use dimension arrays, and some more concise llvm functions.

pcf000 committed:
clang-format

pcf000 committed:
Make a loop to add 0,1,etc key names in maps for MatchLayoutsToInput.

pcf000 committed:
clang-format

pcf000 committed:
Replace 'assert(0)' with report_fatal_error().

pcf000 committed:
clang-format

pcf000 committed:
More assorted 2d-assumption fixes.  Not passing SparseTensor tests.

pcf000 committed:
Handle 1, 2, or 3 dimensions in generateKernel().

pcf000 committed:
Merge branch 'develop' into pf-issue-1311c

pcf000 committed:
Change parseConvDims() to take an array of dimensions instead of separate height/width/(depth).

pcf000 committed:
Merge branch 'develop' into pf-issue-1311c

pcf000 committed:
Corrections after merge.  Fix self-inflicted bug in getConvolutionDims().  Passes check-rocmlir now.

pcf000 committed:
clang-format

pcf000 committed:
Getting a rock.conv3d test case to compile.  Debug code and crude ungeneralised code still present.

pcf000 committed:
Flush debug code, add a self-verifying 3d-rock.conv test.

pcf000 committed:
clang-format

pcf000 committed:
Fix error in changes to MIGraphXToTosa.cpp.

pcf000 committed:
Fix another error in changes to MIGraphXToTosa.cpp.

pcf000 committed:
Account for both 2- and 3-D convolutions in isValid() key check.

pcf000 committed:
clang-format

pcf000 committed:
Remove mfma and atomic_add features from test rock.conv to allow all CI targets.

pcf000 committed:
Changing ConvolutionDims.h to avoid clang-tidy problems.

krzysz00 commented on mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp:
```diff
@@ -213,52 +213,82 @@ LogicalResult ConvConverter<ConvType>::matchAndRewrite(
+                filter.getType().template cast<ShapedType>().getShape()[0],
+                rewriter)});
+    break;
+  default:
```
... Add 1D?

pcf000 replied:
Not sure what to rewrite it into, since there isn't a tosa::Conv1DOp.

krzysz00 commented on mlir/lib/Dialect/Rock/Generator/ConvGenerator.cpp:
```diff
@@ -499,36 +484,35 @@ LogicalResult ConvGenerator::parseConvConfig(OpBuilder &builder,
-                       return argMap.count(key) > 0;
-                     })) {
-      return false;
+    if (argMap["in_layout"].length() > 5) { // Ie, 3-D.
```
Might be nice to have a second vector of keys to `all_of` on for the 3D case but that's me.

krzysz00 reviewed (approved):
I think MIGraphX also wanted 1D convolution, but also,  I can't see any major issues here.

pcf000 committed:
Refactor and generalise correctConvParameters().

pcf000 committed:
clang-format

pcf000 commented:
Replaced by #1505, which has been merged.
